File: backend/src/models/Email.ts
import mongoose, { Schema, Document } from 'mongoose';
import { EMAIL_CATEGORIES } from '../utils/constants';

export interface IEmail extends Document {
  userId: mongoose.Types.ObjectId;
  gmailMessageId: string;
  threadId: string;
  from: string;
  to: string;
  subject: string;
  body: string;
  bodyText: string;
  snippet: string;
  receivedAt: Date;
  category: typeof EMAIL_CATEGORIES[number];
  bank: string;
  isProcessed: boolean;
  hasTransaction: boolean;
  transactionId?: mongoose.Types.ObjectId;
  labels: string[];
  attachments: Array<{
    filename: string;
    mimeType: string;
    size: number;
  }>;
  createdAt: Date;
  updatedAt: Date;
}

const emailSchema = new Schema<IEmail>(
  {
    userId: { type: Schema.Types.ObjectId, ref: 'User', required: true },
    gmailMessageId: { type: String, required: true },
    threadId: String,
    from: { type: String, required: true },
    to: String,
    subject: { type: String, required: true },
    body: String,
    bodyText: String,
    snippet: String,
    receivedAt: { type: Date, required: true },
    category: { type: String, enum: EMAIL_CATEGORIES, default: 'unknown' },
    bank: { type: String, default: 'Unknown' },
    isProcessed: { type: Boolean, default: false },
    hasTransaction: { type: Boolean, default: false },
    transactionId: { type: Schema.Types.ObjectId, ref: 'Transaction' },
    labels: [String],
    attachments: [
      {
        filename: String,
        mimeType: String,
        size: Number,
      },
    ],
  },
  { timestamps: true },
);

emailSchema.index({ userId: 1, gmailMessageId: 1 }, { unique: true });
emailSchema.index({ userId: 1, receivedAt: -1 });
emailSchema.index({ userId: 1, category: 1 });
emailSchema.index({ userId: 1, bank: 1 });
emailSchema.index({ userId: 1, isProcessed: 1 });

export const Email = mongoose.models.Email || mongoose.model<IEmail>('Email', emailSchema);

File: syncflow_app/lib/models/email.dart
class Email {
  final String id;
  final String? gmailMessageId;
  final String? threadId;
  final String from;
  final String? to;
  final String subject;
  final String? snippet;
  final String? body;
  final String? bodyText;
  final String receivedAt;
  final String category;
  final String bank;
  final bool isProcessed;
  final bool hasTransaction;
  final String? transactionId;
  final List<String> labels;
  final String createdAt;

  Email({
    required this.id,
    this.gmailMessageId,
    this.threadId,
    required this.from,
    this.to,
    required this.subject,
    this.snippet,
    this.body,
    this.bodyText,
    required this.receivedAt,
    this.category = 'unknown',
    this.bank = 'Unknown',
    this.isProcessed = false,
    this.hasTransaction = false,
    this.transactionId,
    this.labels = const [],
    required this.createdAt,
  });

  factory Email.fromJson(Map<String, dynamic> json) {
    return Email(
      id: json['_id'] as String? ?? '',
      gmailMessageId: json['gmailMessageId'] as String?,
      threadId: json['threadId'] as String?,
      from: json['from'] as String? ?? '',
      to: json['to'] as String?,
      subject: json['subject'] as String? ?? '',
      snippet: json['snippet'] as String?,
      body: json['body'] as String?,
      bodyText: json['bodyText'] as String?,
      receivedAt: json['receivedAt'] as String? ?? '',
      category: json['category'] as String? ?? 'unknown',
      bank: json['bank'] as String? ?? 'Unknown',
      isProcessed: json['isProcessed'] as bool? ?? false,
      hasTransaction: json['hasTransaction'] as bool? ?? false,
      transactionId: json['transactionId'] as String?,
      labels: (json['labels'] as List<dynamic>?)?.map((e) => e as String).toList() ?? [],
      createdAt: json['createdAt'] as String? ?? '',
    );
  }
}

class EmailStats {
  final int total;
  final int unprocessed;
  final Map<String, int> categories;
  final Map<String, int> banks;

  EmailStats({
    this.total = 0,
    this.unprocessed = 0,
    this.categories = const {},
    this.banks = const {},
  });

  factory EmailStats.fromJson(Map<String, dynamic> json) {
    return EmailStats(
      total: json['total'] as int? ?? 0,
      unprocessed: json['unprocessed'] as int? ?? 0,
      categories: (json['categories'] as Map<String, dynamic>?)
              ?.map((k, v) => MapEntry(k, v as int)) ??
          {},
      banks: (json['banks'] as Map<String, dynamic>?)
              ?.map((k, v) => MapEntry(k, v as int)) ??
          {},
    );
  }
}

class BankInfo {
  final String name;
  final int count;
  final String? lastEmail;

  BankInfo({
    required this.name,
    this.count = 0,
    this.lastEmail,
  });

  factory BankInfo.fromJson(Map<String, dynamic> json) {
    return BankInfo(
      name: json['name'] as String? ?? '',
      count: json['count'] as int? ?? 0,
      lastEmail: json['lastEmail'] as String?,
    );
  }
}

class SyncResult {
  final int? transactionsCreated;
  final int? duplicatesFound;
  final int? sentForReview;
  final List<String>? parseErrors;
  final Map<String, dynamic>? syncResult;

  SyncResult({
    this.transactionsCreated,
    this.duplicatesFound,
    this.sentForReview,
    this.parseErrors,
    this.syncResult,
  });

  factory SyncResult.fromJson(Map<String, dynamic> json) {
    return SyncResult(
      transactionsCreated: json['transactionsCreated'] as int?,
      duplicatesFound: json['duplicatesFound'] as int?,
      sentForReview: json['sentForReview'] as int?,
      parseErrors: (json['parseErrors'] as List<dynamic>?)
          ?.map((e) => e as String)
          .toList(),
      syncResult: json['syncResult'] as Map<String, dynamic>?,
    );
  }
}
